datalake/src/main/java/com/sequenceiq/datalake/flow/upgrade/ccm/UpgradeCcmActions.java

```diff
@@ -0,0 +1,127 @@
+            }
+
+            @Override
+            protected void doExecute(SdxContext context, UpgradeCcmStackEvent payload, Map<Object, Object> variables) {
+                LOGGER.info("Execute CCM upgrade stack flow for SDX: {}", payload.getResourceId());
+                sendEvent(context);
+            }
+
+            @Override
+            protected Selectable createRequest(SdxContext context) {
+                return new UpgradeCcmStackRequest(context.getSdxId(), context.getUserId());
+            }
+
+            @Override
+            protected Object getFailurePayload(UpgradeCcmStackEvent payload, Optional<SdxContext> flowContext, Exception ex) {
+                return UpgradeCcmFailedEvent.from(payload, ex);
+            }
+        };
+    }
+
+    @Bean(name = "UPGRADE_CCM_FINISHED_STATE")
+    public Action<?, ?> finishedAction() {
+        return new AbstractSdxAction<>(UpgradeCcmSuccessEvent.class) {
+            @Override
+            protected SdxContext createFlowContext(FlowParameters flowParameters, StateContext<FlowState, FlowEvent> stateContext,
+                    UpgradeCcmSuccessEvent payload) {
+                return SdxContext.from(flowParameters, payload);
+            }
+
+            @Override
+            protected void doExecute(SdxContext context, UpgradeCcmSuccessEvent payload, Map<Object, Object> variables) {
+                LOGGER.info("CCM upgrade finalized: {}", payload.getResourceId());
+                SdxCluster sdxCluster = sdxStatusService.setStatusForDatalakeAndNotify(DatalakeStatusEnum.RUNNING,
+                        "Datalake is running", payload.getResourceId());
+                metricService.incrementMetricCounter(MetricType.UPGRADE_CCM_FINISHED, sdxCluster);
+                sendEvent(context, UPGRADE_CCM_FINALIZED_EVENT.event(), payload);
+            }
+
+            @Override
+            protected Object getFailurePayload(UpgradeCcmSuccessEvent payload, Optional<SdxContext> flowContext, Exception ex) {
+                return payload;
+            }
+        };
+    }
+
+    @Bean(name = "UPGRADE_CCM_FAILED_STATE")
+    public Action<?, ?> failedAction() {
+        return new AbstractSdxAction<>(UpgradeCcmFailedEvent.class) {
+            @Override
+            protected SdxContext createFlowContext(FlowParameters flowParameters, StateContext<FlowState, FlowEvent> stateContext,
+                    UpgradeCcmFailedEvent payload) {
+                return SdxContext.from(flowParameters, payload);
+            }
+
+            @Override
+            protected void doExecute(SdxContext context, UpgradeCcmFailedEvent payload, Map<Object, Object> variables) {
+                Exception exception = payload.getException();
+                DatalakeStatusEnum failedStatus = DatalakeStatusEnum.DATALAKE_UPGRADE_CCM_FAILED;
+                LOGGER.info("Update SDX status to {} for resource: {}", failedStatus, payload.getResourceId(), exception);
```

Comment:
There is no placeholder for `exception`. Perhaps you meant to pick the `info(String, Throwable)` variant here?